### PR TITLE
EIT-2303: Feature - Cash App Orders

### PR DESCRIPTION
**Pull request by ScottAntonacAP**
## Summary of Changes

- Add methods to sign and validate an Afterpay Cash App Order
- Pass back from the SDK data to be used when creating a Cash App payment

### CODE REVIEW

File: Example/Example/Purchase/CartViewController.swift [hunk 1]
```diff
@@ -19,8 +20,14 @@ final class CartViewController: UIViewController, UITableViewDataSource {
   private let titleSubtitleCellIdentifier = String(describing: TitleSubtitleCell.self)
   private let eventHandler: (Event) -> Void
 
+  private lazy var cashButton = CashAppPayButton(size: .large) {
+    self.didTapCashAppPay()
+  }
```
Comment (mmroz) [inline Example/Example/Purchase/CartViewController.swift:25]:
```suggestion
  private lazy var cashButton = CashAppPayButton(size: .large) { [weak self] in
    self?.didTapCashAppPay()
  }
```

File: Example/Example/Purchase/PurchaseLogicController.swift [hunk 2]
```diff
@@ -118,11 +123,97 @@ final class PurchaseLogicController {
     }
   }
 
+  func payWithCashApp() {
+    if cashRequest != nil {
+      paykit?.authorizeCustomerRequest(cashRequest!)
+    }
```
Comment (mmroz) [inline Example/Example/Purchase/PurchaseLogicController.swift:129]:
```suggestion
    guard let cashRequest else {
      return
    }
    paykit?.authorizeCustomerRequest(cashRequest)
```

File: Example/Example/Purchase/PurchaseLogicController.swift [hunk 3]
```diff
@@ -118,11 +123,97 @@ final class PurchaseLogicController {
...
+  }()
+
+  func enableCashButton() {
+    cashButton?.alpha = 1
+    cashButton?.isEnabled = true
+  }
+
+  func disableCashButton() {
+    cashButton?.isEnabled = false
+    cashButton?.alpha = 0.3
```
Comment (mmroz) [inline Example/Example/Purchase/PurchaseLogicController.swift:158]: We really should provide this in the SDK, I will prompt design on this one but this is fine for now

File: Example/Example/Purchase/PurchaseLogicController.swift [hunk 4]
```diff
@@ -208,3 +299,66 @@ final class PurchaseLogicController {
...
+
+      quantities = [:]
+      commandHandler(.updateProducts(productDisplayModels))
+
+      let cashTag = customerProfile.cashtag
+      let customerId = customerProfile.id
+
+      let formatter = CurrencyFormatter(currencyCode: currencyCode)
+      let total = grants.reduce(0) { $0 + ($1.action.money?.amount ?? 0) }
+      let amount = formatter.string(from: Decimal(total) / 100)
```
Comment (mmroz) [inline Example/Example/Purchase/PurchaseLogicController.swift:342]: Up to you but you can also use string interpolation. For reference: https://www.hackingwithswift.com/articles/178/super-powered-string-interpolation-in-swift-5-0

**Review by mmroz — COMMENTED** [review]
Looks good so far! I will re-review once I am able to get the server running

**Comment by mmroz** [comment]
Would it be possible to post testing instructions to run this against the test server?

**Comment by mmroz** [comment]
How I tested this:
1. Clone the Sample API https://github.com/afterpay/sdk-example-server and follow the read-me instructions
(not included in the readme: replace `AFTERPAY_API_HOSTNAME="global-api-sandbox.afterpay.com"` with `AFTERPAY_API_HOSTNAME="global-api.afterpay.com"` and set `AFTERPAY_REGION="en-US"`)
2. Open the SDK and change the [line](https://github.com/afterpay/sdk-ios/blob/c7ab497b321d358df08cb2a51eaa1b06c2459f69/Example/Example/Shared/Repository.swift#L18) to use `.production`
3. Run the app, add something to the cart and click PayKit Pay button

**Comment by mmroz** [comment]
I noticed that this failed when it redirected to Cash App.

Have you registered the `client ID/ brand ID` you are using in either sandbox or production? 

https://user-images.githubusercontent.com/8743061/220399604-1ea1ff98-af28-4397-8b7b-87dacc60e6b6.mp4

File: Sources/Afterpay/Model/Environment.swift [hunk 5]
```diff
@@ -35,6 +35,33 @@ import Foundation
...
+      return "https://api-plus.us.afterpay.com/v2/payments/validate-payment"
+    }
+  }
+
+  var cashAppClientId: String {
+    switch self {
+    case .sandbox:
+      return "CAS-CI_AFTERPAY"
+    case .production:
+      return "CA-CI_AFTERPAY"
```
Comment (mmroz) [inline Sources/Afterpay/Model/Environment.swift:61]: Same values, but should they be?

File: Example/Example/Purchase/PurchaseLogicController.swift [hunk 6]
```diff
@@ -118,11 +123,97 @@ final class PurchaseLogicController {
...
+      assertionFailure("Couldn't get cash app client id")
+      return nil
+    }
+
+    let paykitSdkEnv = Afterpay.environment == .production ? PayKit.Endpoint.production : PayKit.Endpoint.sandbox
+    let sdk = PayKit(clientID: clientId, endpoint: paykitSdkEnv)
+    sdk.addObserver(self)
+
+    return sdk
+  }()
```
Comment (mmroz) [inline Example/Example/Purchase/PurchaseLogicController.swift:149]: For our demo we only use `sandbox` but maybe you want both? 
Unclear to me when you would want to give our production credentials in this context.

https://cs.github.com/cashapp/cash-app-pay-ios-sdk/blob/cad34a68717756c34505452a407d3e27c770ab7e/Demo/PayKitDemo/PayKitViewController.swift#L49

File: Example/Example/Purchase/PurchaseLogicController.swift [hunk 7]
```diff
@@ -208,3 +299,66 @@ final class PurchaseLogicController {
   }
 
 }
+
+extension PurchaseLogicController: PayKitObserver {
+  func stateDidChange(to state: PayKitState) {
+
+    print("Cash app state change:", Mirror(reflecting: state).children.first!.label!)
```
[the remark this answers is not in the document]
Comment (ScottAntonacAP) [reply Example/Example/Purchase/PurchaseLogicController.swift]: Addressed in [f07f0a8](https://github.com/afterpay/sdk-ios/pull/243/commits/f07f0a87985e2ff7acf6899a83a25c13690ff32b)

File: Example/Example/Shared/APIClient.swift [hunk 8]
```diff
@@ -19,11 +19,13 @@ private struct CheckoutsRequest: Encodable {
   let email: String
   let amount: String
   let mode: String?
+  let isCashAppPay: Bool?
```
[the remark this answers is not in the document]
Comment (ScottAntonacAP) [reply Example/Example/Shared/APIClient.swift]: Updated in [cd9b823](https://github.com/afterpay/sdk-ios/pull/243/commits/cd9b8230cccd39693c09791886dfd1fcabc673e6)

File: Sources/Afterpay/CashApp/CashAppPayCheckout.swift [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    request.httpBody = jsonRequestBody
+
+    return request
+  }
+
+  private func signPayment(
+    request: URLRequest,
+    signingCompletion: @escaping (_ jwt: CashAppSigningResult) -> Void
+  ) {
+    URLSession.shared.dataTask(with: request) { data, response, error in
```
[the remark this answers is not in the document]
Comment (ScottAntonacAP) [reply Sources/Afterpay/CashApp/CashAppPayCheckout.swift]: Added in [9dc870d](https://github.com/afterpay/sdk-ios/pull/243/commits/9dc870ddfd2efeed32aea5f805bb3dc47ae44118)

**Review by mmroz — APPROVED** [review]
This looks great!

File: Example/Example/Purchase/PurchaseLogicController.swift (hunk 3 above)
Comment (mmroz) [reply Example/Example/Purchase/PurchaseLogicController.swift:159]: I have made this change and pushed out a release. Please update to 0.2.0 and you can remove the alpha!

Comment (ScottAntonacAP) [reply Example/Example/Purchase/PurchaseLogicController.swift]: Updated in [ce37b12](https://github.com/afterpay/sdk-ios/pull/243/commits/ce37b12ea106378d96564f214a542848e55aee58)

File: Example/Example/Purchase/PurchaseLogicController.swift [hunk 10]
```diff
@@ -118,11 +123,92 @@ final class PurchaseLogicController {
...
+
+  private func isCashButtonEnabled(_ isEnabled: Bool) {
+    cashButton?.isEnabled = isEnabled
+  }
+
+  func createCashAppRequest(cashButton: CashAppPayButton? = nil) {
+    if cashButton != nil {
+      self.cashButton = cashButton
+    }
+    isCashButtonEnabled(false)
```
Comment (mmroz) [inline Example/Example/Purchase/PurchaseLogicController.swift:160]: Nice 👍

File: Example/Example/Purchase/PurchaseLogicController.swift [hunk 11]
```diff
@@ -194,13 +184,28 @@ final class PurchaseLogicController {
     }
   }
 
-  func loadCashAppToken() {
+  func retrieveCashAppToken(cashButton: CashAppPayButton? = nil) {
+    if cashButton != nil {
+      self.cashButton = cashButton
+    }
+    isCashButtonEnabled(false)
```
Comment (jhankin) [inline Example/Example/Purchase/PurchaseLogicController.swift:191]: This function is incorrectly named -- it should be called `setCashButtonEnabled(_ enabled: Bool)` -- but you could get rid of this function entirely and just rewrite these lines as 

```
if let cashButton = cashButton {
    self.cashButton = cashButton
}
self.cashButton?.isEnabled = false
```

**Review by jhankin — APPROVED** [review]
One comment on naming, otherwise LGTM

File: Example/Example/Purchase/PurchaseLogicController.swift (hunk 11 above)
Comment (ScottAntonacAP) [reply Example/Example/Purchase/PurchaseLogicController.swift]: renamed in [b7f8283](https://github.com/afterpay/sdk-ios/pull/243/commits/b7f8283d641dcf9495116ca216815db7299687af)

**Review by rmarinleal — APPROVED** [review]

**Pull request merged**